aarch64/kernel/nano_core/main.rs

```diff
@@ -30,16 +36,53 @@ fn main(
     uefi_services::init(&mut system_table)
         .map_err(|_| "nano_core::main - couldn't init uefi services")?;
 
-    let bootsvc = system_table.boot_services();
+    let boot_svc = system_table.boot_services();
 
     let safety = 16;
 
-    let mmap_size = bootsvc.memory_map_size();
+    let mmap_size = boot_svc.memory_map_size();
     let mut mmap = vec![0; mmap_size.map_size + safety * mmap_size.entry_size];
 
-    let _ = system_table.exit_boot_services(handle, &mut mmap)
+    let (_runtime_svc, mem_iter) = system_table.exit_boot_services(handle, &mut mmap)
         .map_err(|_| "nano_core::main - couldn't exit uefi boot services")?;
 
+    // Now set up the list of free regions and reserved regions so we can initialize the frame allocator.
+    let mut free_regions: [Option<PhysicalMemoryRegion>; 32] = Default::default();
+    let mut free_index = 0;
+    let mut reserved_regions: [Option<PhysicalMemoryRegion>; 32] = Default::default();
+    let mut reserved_index = 0;
+
+    for descriptor in mem_iter {
+        let size = descriptor.page_count as usize * PAGE_SIZE;
+        if size > 0 {
+            let (dst, index, region_type) = match descriptor.ty {
+                MemoryType::CONVENTIONAL => (&mut free_regions, &mut free_index, MemoryRegionType::Free),
+                _ => (&mut reserved_regions, &mut reserved_index, MemoryRegionType::Reserved),
```

Comment:
There's an important distinction here. The concept of "reserved" memory in the frame allocator is likely different than the concept of what UEFI/bootloaders consider to be "reserved".

In the frame allocator, "reserved" means that it is still available to be used and allocated, but that it is memory dedicated for a special purpose (e.g., device MMIO). Thus, it has to be specifically requested by address. A general allocation request (where no address is requested) cannot allocate from reserved memory.

That doesn't necessarily seem the be the same thing as the `uefi::boot::MemoryType` enum, so I don't think this `_` catch-all match variant is correct. You need to carefully go through each `MemoryType` to see which ones can be considered as "reserved" in Theseus's terminology.

I also think you need to do this for the `free_regions` too, since `MemoryType::CONVENTIONAL` probably isn't the only type of memory that can be considered as free for general purposes by Theseus's frame allocator.

We can discuss this further too in person or on discord if you're unsure about each `MemoryType` variant.